### ISSUE TITLE
Add a BaseList[MyType] feature

### ISSUE THREAD

**Issue by dga-nagra** [pydantic/pydantic#7970]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

Hi,

Pydantic is really oriented toward objects. If we want to validate a list, we have 3 options
* RootModel: My main issue is that we need to use the `root` field. It is not transparent for the end user and cannot be change for a field `data` for example. We can defined __iter__ and __next__ (and __repr__ to hide the field `root`) but this is a boilerplate we need to handle manually
* Use TypeAdapter(List[MyType]) but we must combine it with manual validation function
  ```python
  def check_sorted(...):
      ...
  MyTypeList = TypeAdapter(List[MyType])

  def check_mytype_list(list):
      list = MyTypeList.validate_python(list)
      check_sorted(...)
      return list
  ```
  This works but this is ugly and the final type is not a pydantic BaseModel

It would be great to have a container base class to create a subclass as for and RootModel. The definition would be something like
```python
class MyTypeList(BaseList[int]):
    pass

# Or with a validator
class MyTypeList(BaseList[int]):
    @model_validator(mode='before')
    def check_order(cls, data: List[int]) -> List[int]:
        ...

a = MyTypeList([3, 4, 2, 1])
repr(a)  # MyTypeList([3, 4, 2, 1])
for x in a:
    ...
  ```

and just `BaseList` could be equivalent to  `BaseList[Any]` 

I tried an implementation on top of `RootModel` and it seems pretty okay (It would be great to include an option to `RootModel` to change the name of the field).
What would you think?

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [X] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by sydney-runkle** [comment]
@dga-nagra,

I don't think this is in high enough demand to belong in the library in general, but certainly some workarounds could be included in user code. 

I think you can achieve a lot of what you're looking for with `TypeAdapter(Annotated[List[int], PlainValidator(...)])`.

Closing this as not planned for now. You can read more in the docs here: https://docs.pydantic.dev/latest/concepts/type_adapter/ and here: https://docs.pydantic.dev/latest/concepts/types/#composing-types-via-annotated
